Source/Core/Core/HW/WiimoteEmu/Dynamics.cpp

```diff
@@ -11,47 +11,91 @@
 
-void EmulateShake(NormalizedAccelData* const accel, ControllerEmu::Buttons* const buttons_group,
-                  const double intensity, u8* const shake_step)
+// Given a velocity, acceleration, and maximum jerk value,
+// calculate change in position after a stop in the shortest possible time.
+// Used to smoothly adjust acceleration and come to complete stops at precise positions.
+double CalculateStopDistance(double velocity, double acceleration, double max_jerk)
+{
+  // Math below expects velocity to be non-negative.
+  const auto velocity_flip = (velocity < 0 ? -1 : 1);
+
+  const auto v_0 = velocity * velocity_flip;
+  const auto a_0 = acceleration * velocity_flip;
+  const auto j = max_jerk;
+
+  // Time to reach zero acceleration.
+  const auto t_0 = a_0 / j;
+
+  // Distance to reach zero acceleration.
+  const auto d_0 = std::pow(a_0, 3) / (3 * j * j) + (a_0 * v_0) / j;
+
+  // Velocity at zero acceleration.
+  const auto v_1 = v_0 + a_0 * std::abs(t_0) - std::copysign(j * t_0 * t_0 / 2, t_0);
+
+  // Distance to complete stop.
+  const auto d_1 = std::copysign(std::pow(std::abs(v_1), 3.0 / 2), v_1) / std::sqrt(j);
+
+  return (d_0 + d_1) * velocity_flip;
+}
+
+double CalculateStopDistance(double velocity, double max_accel)
+{
+  return velocity * velocity / (2 * std::copysign(max_accel, velocity));
+}
+
+}  // namespace
+
+namespace WiimoteEmu
+{
+Common::Vec3 EmulateShake(ControllerEmu::Buttons* const buttons_group, const double intensity,
+                          u8* const shake_step)
 {
   // shake is a bitfield of X,Y,Z shake button states
   static const unsigned int btns[] = {0x01, 0x02, 0x04};
   unsigned int shake = 0;
   buttons_group->GetState(&shake, btns);
 
-  for (int i = 0; i != 3; ++i)
+  Common::Vec3 accel;
+
+  for (int i = 0; i != accel.data.size(); ++i)
   {
     if (shake & (1 << i))
     {
-      (&(accel->x))[i] += std::sin(MathUtil::TAU * shake_step[i] / SHAKE_STEP_MAX) * intensity;
+      accel.data[i] = std::sin(MathUtil::TAU * shake_step[i] / SHAKE_STEP_MAX) * intensity *
+                      GRAVITY_ACCELERATION;
       shake_step[i] = (shake_step[i] + 1) % SHAKE_STEP_MAX;
     }
     else
     {
       shake_step[i] = 0;
     }
   }
+
+  return accel;
 }
 
-void EmulateDynamicShake(NormalizedAccelData* const accel, DynamicData& dynamic_data,
-                         ControllerEmu::Buttons* const buttons_group,
-                         const DynamicConfiguration& config, u8* const shake_step)
+Common::Vec3 EmulateDynamicShake(DynamicData& dynamic_data,
+                                 ControllerEmu::Buttons* const buttons_group,
+                                 const DynamicConfiguration& config, u8* const shake_step)
 {
   // shake is a bitfield of X,Y,Z shake button states
   static const unsigned int btns[] = {0x01, 0x02, 0x04};
   unsigned int shake = 0;
   buttons_group->GetState(&shake, btns);
 
+  Common::Vec3 accel;
+
   for (int i = 0; i != 3; ++i)
```

Comment:
This wasn't changed compared to `EmulateShake`, but I guess you also meant this?
```suggestion
   for (size_t i = 0; i != accel.data.size(); ++i)
```

Reply:
:+1: